core/silkworm/common/endian.hpp

```diff
@@ -185,31 +186,33 @@ ByteView to_big_compact(uint64_t value);
 //! \remarks A "compact" big endian form strips leftmost bytes valued to zero
 ByteView to_big_compact(const intx::uint256& value);
 
-//! \brief Parses unsigned integer from a compacted big endian byte form
-//! \param [in] data : byte view of compacted value. Length must be <= sizeof(UnsignedInteger)
-//! \param [in] allow_leading_zeros : when false, return std::nullopt if data starts with a 0 byte (i.e. not compact)
-//! \return The corresponding integer with native endianness; std::nullopt if data is invalid
-//! \remarks A "compact" big endian form strips leftmost bytes valued to zero
+//! \brief Parses unsigned integer from a compacted big endian byte form.
+//! \param [in] data : byte view of compacted value. Length must be <= sizeof(UnsignedInteger);
+//! otherwise kOverflow is returned.
+//! \param [out] out: the corresponding integer with native endianness.
+//! \return kOk or kOverflow or kLeadingZero.
+//! \remarks A "compact" big endian form strips leftmost bytes valued to zero;
+//! if the input is not compact kLeadingZero is returned.
 template <typename UnsignedInteger>
-static std::optional<UnsignedInteger> from_big_compact(ByteView data, bool allow_leading_zeros = false) {
+static DecodingResult from_big_compact(ByteView data, UnsignedInteger& out) {
```

Comment:
Why remove the possibility to "allow_leading_zeroes" ? I can't recall the rationale but I guess there was a reason for it to be there.

Reply:
Also shouldn't we specialize the template ?
```
template <typename UnsignedInteger,
          typename std::enable_if<std::is_integral<UnsignedInteger>::value && std::is_unsigned<UnsignedInteger>::value, T>::type* = nullptr>
```

Reply:
Unfortunately `is_integral` is not true for `intx`. Perhaps @chfast can enable it.

Reply:
> Why remove the possibility to "allow_leading_zeroes" ? I can't recall the rationale but I guess there was a reason for it to be there.

I dropped allow_leading_zeroes for simplicity. It was used only in one place in cmd/consensus, where I replaced it with `std::stoull` base=16).

Reply:
> Unfortunately `is_integral` is not true for `intx`. Perhaps @chfast can enable it.

The `std::is_intergral` is not user-extendable. https://en.cppreference.com/w/cpp/types/is_integral